### ISSUE TITLE
[R-Forge #2605] add filtering option to fread so it can load less than all rows

### ISSUE THREAD

**Issue by arunsrinivasan** [Rdatatable/data.table#583]
Submitted by: stat quant; Assigned to: Nobody; [R-Forge link](https://r-forge.r-project.org/tracker/index.php?func=detail&aid=2605&group_id=240&atid=978)

Discussed in [data.table list](http://r.789695.n4.nabble.com/fread-suggestion-tp4660924.html).

`fread(input,  chunk.nrows=10000, chunk.filter = <anything acceptable to i of DT[i]>)`, that <anything> could be `grep()` or any expression of column names.

**Comment by MichaelChirico** [comment]
I'm pretty sure this is the same as what I had in mind recently, but let me elaborate with an example:

    read_dt<-data.table(id=sample(10,1e7,T),var=rnorm(1e7))
    write.csv(read_dt,file="dt_to_read.csv",quote=F,row.names=F)
    main_dt<-data.table(id=sample(8,1e5,T),var2=rnorm(1e5))

I'm working with `main_dt` but want to pull in some matching (based on `id`) info from `read_dt`; currently, I need to do something like this:

    relevant_read_dt<-fread("dt_to_read.csv")[id %in% main_dt[,unique(id)]]

This is inefficient because I need to read all of `read_dt` (especially painful as the number of rows of `read_dt` increases, then immediately chop off ~20% of them.

An approach like this:

    relevant_read_dt<-fread("dt_to_read.csv",row.select=id %in% main_dt[,unique(id)])

Would only require 1) read `id` from `"dt_to_read.csv"` 2) run the logical argument `id %in% main_dt[,unique(id)]` and return row numbers to read 3) `fread` only the selected row numbers.

**Comment by skanskan** [comment]
Hello.
Is the filtering already implemented?
For example I want to read a very big csv file  with 4 columns: Value, XXX, YYY, ZZZ, 
and I want to read only the lines where  the  Value >= 1.3

I could do it in two steps: first read all the file, second filter,  but this is slower and I could have problems if the file doesn't fit on memory.

I don't know if we are speaking about the same thing or if I missunderstood it.

Regards.

**Comment by jangorecki** [comment]
@skanskan Yes, this ticket is about exactly that. No, it wasn't yet implemented.

**Comment by skanskan** [comment]
OK, thank you @jangorecki

**Comment by zachmayer** [comment]
I was wondering if any progress has been made here?  It'd be a super useful feature!

**Comment by VinceLYO** [comment]
Any update for those stuck with Windows :D ?

**Comment by VinceLYO** [comment]
**UPDATE 2 :**

After some tests, I figured that the solution I proposed wasn't working on R.
Actually, the code line `tail -n+2 fichier.txt | shuf -n 15`  works in a cmd consol, but not in R with Fread.
It returns the header as a line (randomly, of course).

This issue can be reproduced with the[ iris dataset ](https://fr.osdn.net/projects/sfnet_irisdss/downloads/IRIS.csv/) and the following code :
```
setwd("path")
test <- fread("tail -n+2 IRIS.csv | shuf -n 149"
              ,sep = ","
              ,header = FALSE)
```

You can also try with `sed 1d IRIS.csv | shuf -n149` => Same result.

Does fread deal with pipe and command lines more complicated than one instruction ?

Thanks

Vincent.

**Comment by skanskan** [comment]
@VinceLYO  and all this seems nice but have a problem, it depends on external commands not available on all computers.  It would be great to have a self-contained solution using just fread or maybe R commands.

**Comment by VinceLYO** [comment]
@skanskan : I do agree ! I think it's a nice feature. Let's hope for this :)

**Comment by MichaelChirico** [comment]
To be updated

https://stackoverflow.com/questions/47172355

**Comment by deepakagrawal** [comment]
Any updates on this feature? Is anyone working on this?

**Comment by woodywonge** [comment]
Any updates on this feature?

**Comment by woodywonge** [comment]
Any updates on this feature?

**Comment by tgwhite** [comment]
Boop

**Comment by jangorecki** [comment]
For those who are bumping this issue, be sure to upvote first post here as well. I will clean up a little bit this thread.
----
Regarding the FR itself. I don't think it make sense to introduce new mechanism for filtering on a csv files directly. It is basically a lot of effort and maintenance, where now `grep` works pretty well. What could eventually be a low hanging fruit, is to examine filter expression, guess which columns are required to filter. Then read fully those columns only, perform filter using currently implemented algorithms `which=TRUE`, and then re-read csv applying filter on lines based on `which` results. That would be fully implemented in R, might not be so efficient, but should reduce peak memory required.

**Comment by MichaelChirico** [comment]
See here: https://stackoverflow.com/a/62240442/3576984

`grep` / `awk` don't have the benefit of autoparallelism so can be quite slow vs `fread`

**Comment by MichaelChirico** [comment]
> `grep` works pretty well

One issue I don't see raised yet that's a shortcoming of many `sys` + `fread` approaches is that the first row may be lost, so we won't get nice column names unless we're extra careful, e.g.

```
fwrite(as.data.table(mtcars, keep.rownames="name"), tmp <- tempfile())
fread(paste("grep -F 'Merc'", tmp))
#             V1   V2 V3    V4  V5   V6   V7   V8 V9 V10 V11 V12
# 1:   Merc 240D 24.4  4 146.7  62 3.69 3.19 20.0  1   0   4   2
# 2:    Merc 230 22.8  4 140.8  95 3.92 3.15 22.9  1   0   4   2
# 3:    Merc 280 19.2  6 167.6 123 3.92 3.44 18.3  1   0   4   4
# 4:   Merc 280C 17.8  6 167.6 123 3.92 3.44 18.9  1   0   4   4
# 5:  Merc 450SE 16.4  8 275.8 180 3.07 4.07 17.4  0   0   3   3
# 6:  Merc 450SL 17.3  8 275.8 180 3.07 3.73 17.6  0   0   3   3
# 7: Merc 450SLC 15.2  8 275.8 180 3.07 3.78 18.0  0   0   3   3
```

Would it be worth adding an argument to `fread` that would work around this somehow? That would surely require a lot less development work than filtering. Mostly a question of design.

**Comment by MichaelChirico** [comment]
How about using `col.names = /path/to/file` for this?

The only overlap with current usage is for one-column files; it should be safe to check `file.exists(col.names)` to distinguish the two cases.

**Comment by luisvalenzuelar** [comment]
It's interesting that neither Python nor Stata nor other R functions like readr's `read_csv` have managed to include this option. In any case, [benchmarking](https://stackoverflow.com/a/40335203/3585386) suggests that loading all data before subsetting using commands like `read_csv_chunked` or `read.csv.sql` do better than system-based approaches (grwp/awk/etc), approaches which are in any case far from intuitive. Maybe `fread` can allow for the pre-loading options, which might still be faster than subsetting ex-post, i.e. `A[B]`.
